### PR TITLE
Support join domains

### PR DESCRIPTION
**Pull request by mikumikuchan**
Based on https://github.com/vijos/vj4/pull/353, https://github.com/vijos/vj4/pull/354

### CODE REVIEW

File: vj4/error.py [hunk 1]
```diff
@@ -230,6 +230,24 @@ def message(self):
...
+    return 'You are not allowed to join the domain. The link is either invalid or expired.'
+
+
+class UserAlreadyDomainMemberError(ForbiddenError):
+  @property
+  def message(self):
+    return 'You are already a member of the domain.'
+
+
+class InvalidJoinInvitationCodeError(ForbiddenError):
```
Comment (twd2) [inline vj4/error.py]: `InvalidInvitationCodeError`

File: vj4/handler/domain.py [hunk 2]
```diff
@@ -93,6 +97,86 @@ class DomainEditHandler(base.Handler):
...
+        join_settings['expire'] = datetime.datetime.utcnow() + datetime.timedelta(hours=expire)
+    await domain.edit(self.domain_id, join=join_settings)
+    self.json_or_redirect(self.url)
+
+
+@app.route('/join', 'domain_join')
+class DomainJoinHandler(base.Handler):
+  async def ensure_user_not_member(self):
+    dudoc = await domain.get_user(self.domain_id, self.user['_id'])
+    if dudoc != None and 'role' in dudoc:
```
Comment (twd2) [inline vj4/handler/domain.py]: `if dudoc and 'role'`...

File: vj4/handler/domain.py [hunk 3]
```diff
@@ -93,6 +97,86 @@ class DomainEditHandler(base.Handler):
...
+      raise error.UserAlreadyDomainMemberError(self.domain_id, self.user['_id'])
+
+  @base.require_priv(builtin.PRIV_USER_PROFILE)
+  @base.get_argument
+  @base.sanitize
+  async def get(self, *, code: str=''):
+    join_settings = domain.get_join_settings(self.domain)
+    if join_settings is None:
```
Comment (twd2) [inline vj4/handler/domain.py]: `if not`

File: vj4/handler/domain.py [hunk 4]
```diff
@@ -93,6 +97,86 @@ class DomainEditHandler(base.Handler):
...
+      raise error.DomainJoinForbiddenError(self.domain_id)
+    await self.ensure_user_not_member()
```
Comment (twd2) [inline vj4/handler/domain.py]: Check this before join_settings

File: vj4/handler/domain.py [hunk 5]
```diff
@@ -93,6 +97,86 @@ class DomainEditHandler(base.Handler):
...
+    self.render('domain_join.html', join_settings=join_settings, code=code)
+
+  @base.require_priv(builtin.PRIV_USER_PROFILE)
+  @base.post_argument
+  @base.require_csrf_token
+  @base.sanitize
+  async def post(self, *, code: str=''):
+    join_settings = domain.get_join_settings(self.domain)
+    if join_settings is None:
```
Comment (twd2) [inline vj4/handler/domain.py]: ditto

File: vj4/handler/domain.py [hunk 6]
```diff
@@ -93,6 +97,86 @@ class DomainEditHandler(base.Handler):
...
+      raise error.DomainJoinForbiddenError(self.domain_id)
+    await self.ensure_user_not_member()
```
Comment (twd2) [inline vj4/handler/domain.py]: ditto

File: vj4/handler/domain.py [hunk 7]
```diff
@@ -93,6 +97,86 @@ class DomainEditHandler(base.Handler):
...
+    if join_settings['method'] == constant.domain.JOIN_METHOD_CODE:
+      if join_settings['code'] != code:
+        raise error.InvalidJoinInvitationCodeError(self.domain_id)
+    # TODO: should be replaced by domain.add_user_role
```
Comment (twd2) [inline vj4/handler/domain.py]: ???

File: vj4/handler/domain.py (hunk 4 above)
Comment (breeswish) [reply vj4/handler/domain.py]: Seems that they are similar. both make sense.

**Pull request merged**